### PR TITLE
Musig2 Implementation

### PR DESCRIPTION
**Pull request by nkohen**
Implements https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki

TODO:
- [x] Add missing safety checks
- [x] Add static test vectors
- [x] Implement and test partial signature verification
- [x] Test with more signers
- [x] Implement tweaking support
- [x] Refactor types
- [x] Support js
- [ ] Test against secp256k1-zkp implementation of MuSig2

Edit: Probably won't bind to and test against the secp256k1-zkp implementation in this PR

### CODE REVIEW

File: crypto/src/main/scala/org/bitcoins/crypto/MuSig2Util.scala [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+  def genMultiNonce(
+      privKeyOpt: Option[ECPrivateKey] = None,
+      aggPubKeyOpt: Option[SchnorrPublicKey] = None,
+      msgOpt: Option[ByteVector] = None,
+      extraInOpt: Option[ByteVector] = None): (
+      MultiNoncePub,
+      MultiNoncePriv) = {
+    msgOpt.foreach(msg => require(msg.length == 32))
+    extraInOpt.foreach(extraIn => require(extraIn.length <= 4294967295L))
```
Comment (benthecarman) [inline crypto/src/main/scala/org/bitcoins/crypto/MuSig2Util.scala]: nit: more readable imo if you do `require(x.forAll(...))`

File: crypto/src/main/scala/org/bitcoins/crypto/MuSig2Util.scala [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+    val preRand =
+      ECPrivateKey.freshPrivateKey.bytes // TODO this needs to draw 32 bytes uniformily at random
```
Comment (benthecarman) [inline crypto/src/main/scala/org/bitcoins/crypto/MuSig2Util.scala]: is that not what this does?

File: crypto/src/main/scala/org/bitcoins/crypto/MuSig2Util.scala [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+      gacc: Boolean) {
+    def getPubKey: ByteVector = aggPubKey.bytes.tail
+    def getPubKey33: ByteVector = aggPubKey.bytes
+  }
+
+  case class KeySet(keys: Vector[SchnorrPublicKey]) {
+    keys.init.zip(keys.tail).foreach { case (key1, key2) =>
+      require(key1.hex.compareTo(key2.hex) <= 0,
+              "Keys must be sorted lexicographically")
+    }
```
Comment (benthecarman) [inline crypto/src/main/scala/org/bitcoins/crypto/MuSig2Util.scala]: might be eaiser to make a `Ordering[SchnorrPublicKey]` so you can just call `.sorted`

File: crypto/src/main/scala/org/bitcoins/crypto/MuSig2Util.scala (hunk 2 above)
Comment (nkohen) [reply crypto/src/main/scala/org/bitcoins/crypto/MuSig2Util.scala]: If I'm not mistaken it almost is, but this throws away a random draw if it is bigger than the field order (which is just slightly less than 2^32), so I need to rewrite without the throw away code

File: crypto/src/main/scala/org/bitcoins/crypto/MuSig2Util.scala (hunk 3 above)
Comment (nkohen) [reply crypto/src/main/scala/org/bitcoins/crypto/MuSig2Util.scala]: Perhaps we should just have a lexicographic ordering defined for all NetworkElements. Now that I say this I vaguely recall already doing this but maybe it's false deja vu, I'll look for it

File: crypto/src/main/scala/org/bitcoins/crypto/MuSig2Util.scala (hunk 2 above)
Comment (nkohen) [reply crypto/src/main/scala/org/bitcoins/crypto/MuSig2Util.scala]: Turns out we already have `CryptoUtil.randomBytes`, so I use this :)

File: crypto/src/main/scala/org/bitcoins/crypto/MuSig2Util.scala [hunk 4]
```diff
@@ -92,21 +81,35 @@ object MuSig2Util {
         .reduce(_.add(_))
     }
 
-    lazy val secondKey: SchnorrPublicKey = keys.tail.head
+    // In truth this represents the first key different from the head key
+    lazy val secondKeyOpt: Option[SchnorrPublicKey] = {
+      keys.find(_ != keys.head)
```
Comment (benthecarman) [inline crypto/src/main/scala/org/bitcoins/crypto/MuSig2Util.scala]: why not do `keys.tail.headOption`

Comment (nkohen) [reply crypto/src/main/scala/org/bitcoins/crypto/MuSig2Util.scala]: As I discovered while implementing test vectors, if the same pubkey is used in the 0,1 spots then you have to take the next one

File: crypto/src/main/scala/org/bitcoins/crypto/CryptoRuntime.scala [hunk 5]
```diff
@@ -211,6 +211,17 @@ trait CryptoRuntime {
 
   def tweakMultiply(publicKey: ECPublicKey, tweak: FieldElement): ECPublicKey
 
+  def tweakMultiply(point: SecpPoint, tweak: FieldElement): SecpPoint = {
+    point match {
+      case SecpPointInfinity => SecpPointInfinity
+      case pt: SecpPointFinite =>
+        Try(tweakMultiply(pt.toPublicKey, tweak)) match {
+          case Failure(_)       => SecpPointInfinity
```
Comment (nkohen) [inline crypto/src/main/scala/org/bitcoins/crypto/CryptoRuntime.scala]: This really should double check by multiplying by tweak.subtract(Field element.one) and checking they're opposite

File: crypto/src/main/scala/org/bitcoins/crypto/CryptoBytesUtil.scala [hunk 6]
```diff
@@ -99,6 +99,18 @@ trait CryptoBytesUtil {
   final def toByteVector[T <: NetworkElement](h: Seq[T]): ByteVector = {
     ByteVector.concat(h.map(_.bytes))
   }
+
+  def splitEvery(bytes: ByteVector, intervalLen: Int): Vector[ByteVector] = {
```
Comment (benthecarman) [inline crypto/src/main/scala/org/bitcoins/crypto/CryptoBytesUtil.scala]: this already exists, call `.grouped`

File: crypto/src/main/scala/org/bitcoins/crypto/musig/MuSigNoncePub.scala [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+      case SecpPointInfinity  => CryptoParams.getG
+      case p: SecpPointFinite => p.toPublicKey
+    }
+  }
+}
+
+object MuSigNoncePub extends Factory[MuSigNoncePub] {
+
+  /** In the BIP, the point at infinity is serialized as 33 0x00 bytes */
+  val infPtBytes: ByteVector = ByteVector.fill(33)(0)
```
Comment (benthecarman) [inline crypto/src/main/scala/org/bitcoins/crypto/musig/MuSigNoncePub.scala]: nit: `ByteVector.low(33)`

**Review by Christewart — COMMENTED** [review]
Can you add a `musig.md` or something with an example?

**Review by Christewart — APPROVED** [review]

**Pull request merged**